### PR TITLE
[CUDA] Fix float32 std/var M2 overflow by using double accumulation

### LINKED ISSUE #180156: GPU two-pass variance: intermediate squared values overflow float32. CPU Welford online algorithm avoids overflow.
### 🐛 Describe the bug

```
import torch

torch.manual_seed(0)
x = torch.randn(1000, dtype=torch.float32) * 1e19 + 1e20

ref = torch.std(x.double()).item()
cpu = torch.std(x).item()
gpu = torch.std(x.cuda()).cpu().item()

print(f"Reference (float64): {ref:.4e}")
print(f"CPU (float32):       {cpu:.4e}")
print(f"GPU (float32):       {gpu}   <-- {'BUG' if not __import__('math').isinf(cpu) and __import__('math').isinf(gpu) else 'ok'}")
```

### Versions

Reference (float64): 1.0287e+19
CPU (float32):       1.0287e+19
GPU (float32):       inf   <-- BUG

cc @ptrblck @msaroufim @eqy @jerryzh168 @tinglvv @nWEIdia

### PR DESCRIPTION
**Pull request by senekis316**
## Summary

`torch.std` and `torch.var` return `inf` on CUDA for `float32` inputs
with large magnitudes, while CPU returns the correct result.

```python
import torch

torch.manual_seed(0)
x = torch.randn(1000, dtype=torch.float32) * 1e19 + 1e20

print(torch.std(x.double()))   # 1.0287e+19  (correct)
print(torch.std(x))            # 1.0287e+19  (correct, CPU)
print(torch.std(x.cuda()))     # inf          (overflow, CUDA)

Root cause: std_var_kernel_impl in ReduceMomentKernel.cu uses
at::acc_type<scalar_t, true> for the Welford accumulation type, which
resolves to float for float32 inputs on CUDA. The intermediate M2
value overflows float32's max (~3.4e38) when:

M2 ~ N × std² ~ 1000 × (1e19)² = 1e41  >  3.4e38
The CPU kernel explicitly uses double for accumulation and is unaffected.

Fix
Make accscalar_t an overridable template parameter in
std_var_kernel_impl (keeping at::acc_type<scalar_t, true> as the
default), and add an explicit float32 → double dispatch branch in
std_var_kernel_cuda so that float32 inputs always accumulate M2 in
double precision.

The Intel XPU team applied the same approach for XPU in
[intel/torch-xpu-ops#3337](https://github.com/intel/torch-xpu-ops/pull/3337).

Test
Added test_std_var_large_float32_no_overflow in test/test_reductions.py,
which covers:

1D global reduce (std and var)
2D reduce along a dimension (exercises parallel combine())
correction=0 (biased variance)
Fixes #180156

### CODE REVIEW

**Comment by Skylion007** [comment]
Don't we need some benchmarks to show this didn't slow things down too much?

**Review by eqy — COMMENTED** [review]
`x = torch.randn(1000, dtype=torch.float32) * 1e19 + 1e20` Is this a realistic distribution or a synthetic case?

**Comment by senekis316** [comment]
> Don't we need some benchmarks to show this didn't slow things down too much?

Hi, I've run a performance benchmark to measure the overhead introduced by this fix. Here are the results:

**Environment:**
- GPU: NVIDIA GeForce RTX 2060 (sm_75)
- CUDA: 12.4
- Baseline: PyTorch 2.13.0a0+git27b9a62 (unpatched main, float32 accumulation)
- Fixed: PyTorch 2.13.0a0+gitce04723 (this branch, double accumulation for float32 inputs)

| size | f32 old (baseline) | f32 fix (this PR) | f64 old (upper bound) | fix/old |
|------|--------------------|-------------------|-----------------------|---------|
| 10,000 | 0.010 ms | 0.074 ms | 0.067 ms | 7.4x |
| 100,000 | 0.030 ms | 0.473 ms | 0.434 ms | 15.8x |
| 1,000,000 | 0.025 ms | 0.208 ms | 0.196 ms | 8.3x |
| 10,000,000 | 0.180 ms | 1.822 ms | 1.689 ms | 10.1x |

The fixed f32 performance is very close to the f64 baseline, which makes sense — the fix internally promotes float32 accumulation to double precision.

One thing I noticed: f32 fix is slightly slower than f64 old, even though f32 fix reads half the data (float32 vs float64). My guess is that the per-element cast from float32 → float64 offsets the memory bandwidth advantage.

I'd appreciate your guidance on whether this overhead is acceptable. 
I can see two potential follow-up optimizations if needed:

1. **Conditional promotion**: add a fast pre-scan (max-reduction) to check 
   whether `max(|x|)` exceeds ~1e19. Only promote to double when needed, 
   preserving the float32 fast path for typical inputs.

2. **Mixed-precision Welford**: keep `mean` and `delta` as float32, but 
   accumulate only `M2` in double. This reduces double-precision operations 
   from 5 to 1 per element (~1/5 of the current overhead), at the cost of 
   more complex `WelfordData` templating. Note that per-element 
   `delta * new_delta` products could still overflow float32 for very large 
   inputs (>~1.8e19), so this approach would need a tighter threshold check.

Happy to explore either direction if the team considers the current overhead 
unacceptable for the correctness fix.

**Comment by senekis316** [comment]
> `x = torch.randn(1000, dtype=torch.float32) * 1e19 + 1e20` Is this a realistic distribution or a synthetic case?

The test input is synthetic — designed to reliably reproduce the overflow —
but the underlying bug was reported by a real user.
Large-magnitude float32 data can arise in scientific computing, financial
modeling, or training pipelines where inputs are not yet normalized.
The synthetic test ensures the regression case is reproducible and stable in CI.
